dash/development/base_component.py

```diff
@@ -387,6 +387,120 @@ def __repr__(self):
     required_args = required_props(props)
     return c.format(**locals())
 
+def generate_class_string_r(typename, props, description, namespace):
+    """
+    Dynamically generate class strings to have nicely formatted docstrings,
+    keyword arguments, and repr
+
+    Inspired by http://jameso.be/2013/08/06/namedtuple.html
+
+    Parameters
+    ----------
+    typename
+    props
+    description
+    namespace
+
+    Returns
+    -------
+    string
+
+    """
+    # TODO _prop_names, _type, _namespace, available_events,
+    # and available_properties
+    # can be modified by a Dash JS developer via setattr
+    # TODO - Tab out the repr for the repr of these components to make it
+    # look more like a hierarchical tree
+    # TODO - Include "description" "defaultValue" in the repr and docstring
+    #
+    # TODO - Handle "required"
+    #
+    # TODO - How to handle user-given `null` values? I want to include
+    # an expanded docstring like Dropdown(value=None, id=None)
+    # but by templating in those None values, I have no way of knowing
+    # whether a property is None because the user explicitly wanted
+    # it to be `null` or whether that was just the default value.
+    # The solution might be to deal with default values better although
+    # not all component authors will supply those.
+    c = '''{docstring}
+
+    html{typename} <- function(..., {default_argtext}) {{
+
+    component <- list(
+      props = list(
+         {default_paramtext}
+      ),
+      type = '{typename}',
+      namespace = '{namespace}',
+      propNames = c({list_of_valid_keys}),
+      package = '{package_name}'
+    )
+
+    component$props <- filter_null(component$props)
+    component <- append_wildcard_props(component, wildcards = c({default_wildcards}), ...)
+
+    structure(component, class = c('dash_component', 'list'))
+    }}
+'''
+
+    filtered_props = reorder_props(filter_props(props))
+    # pylint: disable=unused-variable
+    list_of_valid_wildcard_attr_prefixes = repr(parse_wildcards(props))
+    # pylint: disable=unused-variable
+    list_of_valid_keys = repr(list(map(str, filtered_props.keys())))
+    list_of_valid_keys = list_of_valid_keys[1:-1]
+    package_name = make_package_name(namespace)
+    # pylint: disable=unused-variable
+    docstring = create_docstring_r(
+        component_name=typename,
+        props=filtered_props,
+        events=parse_events(props),
+        description=description).replace('\r\n', '\n')
```

Comment:
I added that `.replace('\r\n', '\n')` for generating the docstring on windows because git doesn't auto replace them when they are in docstring. Not sure this problem is the same with R.

Reply:
Actually, it is a consideration in R as well, though I think I should remove the `\n` and replace with a single space, since R's help viewer does its own line wrapping without adding linefeeds.